File: src/704 Binary Search/main.py
# Solved by @kitanoyoru
# https://leetcode.com/problems/binary-search/

from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        [start, end] = [0, len(nums) - 1]

        while start <= end:
            mid = start + ((end - start) >> 1)
            if nums[mid] == target:
                return mid
            elif nums[mid] > target:
                end = mid - 1
            else:
                start = mid + 1

        return -1

File: src/704 Binary Search/main.ex
defmodule Solution do
  @spec search(nums :: [integer], target :: integer) :: integer
  def search(nums, target), do: search(nums, target, 0, length(nums) - 1)
  def search(_nums, _target, left, right) when left > right, do: -1
  def search(nums, target, left, right) do
    mid_pos = left + div(right - left, 2)
    mid_elem = Enum.at(nums, mid_pos)

    cond do
      mid_elem == target -> mid_pos
      mid_elem < target -> search(nums, target, mid_pos + 1, right)
      mid_elem > target -> search(nums, target, left, mid_pos - 1)
    end
  end
end
